cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationsDashboardHttpHandler.java

```diff
@@ -67,30 +69,25 @@ public OperationsDashboardHttpHandler(Store store) {
   public void readDashboardDetail(FullHttpRequest request, HttpResponder responder,
                                   @QueryParam("start") long startTimeSecs,
                                   @QueryParam("duration") int durationTimeSecs,
-                                  @QueryParam("namespace") Set<String> namespaces)
-    throws IOException, BadRequestException {
+                                  @QueryParam("namespace") Set<String> namespaces) throws BadRequestException {
     if (startTimeSecs < 0) {
       throw new BadRequestException("'start' time cannot be smaller than 0.");
     }
     if (durationTimeSecs < 0) {
       throw new BadRequestException("'duration' cannot be smaller than 0.");
     }
 
-    Set<NamespaceId> namespaceIds = namespaces.stream().map(NamespaceId::new).collect(Collectors.toSet());
-    Map<ProgramRunId, RunRecordMeta> historicalRuns =
-      // TODO: [CDAP-13352] Currently, to get active program runs within a time range,
-      // a full table scan is required in AppMetaStore. Performance improvement will be done.
-      store.getHistoricalRuns(namespaceIds, startTimeSecs, startTimeSecs + durationTimeSecs, Integer.MAX_VALUE);
-    // get historical runs within the query time range
-    Stream<DashboardProgramRunRecord> historicalRecords =
-      historicalRuns.values().stream().map(OperationsDashboardHttpHandler::runRecordToDashboardRecord);
-    // get active runs with start time earlier than the end of query time range
-    Stream<DashboardProgramRunRecord> activeRecords =
-      store.getActiveRuns(namespaceIds, run -> run.getStartTs() < startTimeSecs + durationTimeSecs).values().stream()
-      .map(OperationsDashboardHttpHandler::runRecordToDashboardRecord);
-    responder.sendJson(HttpResponseStatus.OK,
-                       // combine historical runs and active runs
-                       GSON.toJson(Stream.concat(historicalRecords, activeRecords).collect(Collectors.toSet())));
+    byte[] startRow = Bytes.toBytes(startTimeSecs);
+    // endRow is exclusive, adding one
+    byte[] endRow = Bytes.toBytes(startTimeSecs + durationTimeSecs + 1);
+
+    Collection<RunRecordMeta> runRecordMetas = programHeartbeatService.scan(startRow, endRow);
```

Comment:
instead of getting this collection then filtering it below, I think it would be good to support a scan method that takes a `Predicate<RunRecordMeta>` so that we use less memory and can do it in a single pass.